New Recipes + Unit Tests + Bug Fix | 3 September 2024

AliSoltanianFJ opened the pull request:
## List of changes in this PR:
### New Recipes:

- Added recipes to convert `BinaryData`
    - Added a recipe to change instances of `TypeReference` in `toObject()` to `ParameterizedType`
    - Added a recipe to convert `com.azure.core.util.BinaryData` to `io.clientcore.core.util.binarydata.BinaryData`

- Added recipe to convert `com.azure.core.util.Configuration` to `io.clientcore.core.util.configuration.Configuration`

- Added recipe to convert `com.azure.core.util.logging.ClientLogger` to `io.clientcore.core.util.ClientLogger`

- Added recipes to convert `ClientAuthenticationException`, `HttpResponseException`, `ResourceModifiedException` and `ResourceNotFoundException`

- Added recipe to convert `com.azure.core.credential` package name to `io.clientcore.core.credential`


### Other Changes and Bug Fixes:
- Fixed a bug in the `RetryOptionsConstructorRecipe` that would cause it to crash if it encountered an initializer without `new` keyword
- Added unit tests for all of the new recipes

### Resolved Issues:
resolves #7
resolves #1 

### Discarded Changes:
- ~Added recipes to convert `@ UnexpectedResponseExceptionType` annotations and its attributes~
    - ~Added recipe to change annotation attribute names~
    - ~Added recipe to convert `com.azure.core.annotation.UnexpectedResponseExceptionType` to `io.clientcore.core.http.annotation.UnexpectedResponseExceptionDetail`~
    - ~Added recipe to convert given attribute values to azure core v2 naming convention. For example, `ClientAuthenticationException.class` is changed to `CLIENT_AUTHENTICATION`~

- ~Added recipes to convert `@ HostParam`, `@ HeaderParam`, `@ BodyParam`, and `@ QueryParam` annotations~

-~Added recipes to convert `@ ServiceClient`, `@ ServiceClientBuilder`, `@ ServiceInterface`, and `@ ServiceMethod` annotations~

- ~Added recipes to convert `@ Generated`, `@ Immutable`, and `@ ReturnType` annotations~

Casperdeww committed:
Rename Recipe First Commit

This is the first commit for the Rename Recipe, issue should be solved however testing may show errors

AliSoltanianFJ committed:
Merge pull request #29 from AliSoltanianFJ/AddUnitTests14082024

Added unit tests and combined HttpRetry recipes

jessicalangg committed:
Added ChangeType recipes for Param Annotation Types.

jessicalangg committed:
Added ChangeType recipes for Service Annotation Types.

AliSoltanianFJ committed:
Merge pull request #37 from AliSoltanianFJ/Issue33ParamAnnotationTypes

Issue 33 Param Annotation Types

AliSoltanianFJ committed:
Merge branch 'main' into Issue34ServiceAnnotationTypes

AliSoltanianFJ committed:
Merge pull request #38 from AliSoltanianFJ/Issue34ServiceAnnotationTypes

Issue 34 Service Annotation Types.

Casperdeww committed:
Updated yaml

AliSoltanianFJ committed:
Remove redundant file

AliSoltanianFJ committed:
Fix recipe type

AliSoltanianFJ committed:
Merge branch 'main' into Issue2

AliSoltanianFJ committed:
Merge pull request #39 from AliSoltanianFJ/Issue2

Change package name for "com.azure.core.credential" to "io.clientcore.core.credential"

AliSoltanianFJ committed:
Added recipes for miscellaneous Annotation types and new unit test

AliSoltanianFJ committed:
Added unit tests for @-Param Annotation Types

AliSoltanianFJ committed:
Merge pull request #41 from AliSoltanianFJ/AddAnnotationTypeRecipesAndUnitTests

Add annotation type recipes and unit tests

AliSoltanianFJ committed:
Added dependencies for building Maven project

AliSoltanianFJ committed:
Merge pull request #42 from AliSoltanianFJ/AddDependencies

Added dependencies for building Maven project

stoatus-quo committed:
Unit tests added for HttpTrait ChangeMethodName recipes

stoatus-quo committed:
Unit tests added for Annotation ChangeType recipes

stoatus-quo committed:
HttpTraitTest tidied-up

stoatus-quo committed:
Merge pull request #45 from AliSoltanianFJ/unit-tests-backlog

Unit tests backlog

AliSoltanianFJ committed:
Sync with client repo

AliSoltanianFJ committed:
Merge pull request #46 from AliSoltanianFJ/SyncWithClientRepo23082024

Sync with Client Repository

AliSoltanianFJ committed:
Refactored unit tests according to client feedback for previous PR

AliSoltanianFJ committed:
Merge pull request #47 from AliSoltanianFJ/RefactorUnitTests23082024

Refactored unit tests

AliSoltanianFJ committed:
Fixed bug that crashed the RetryOptionsConstructorRecipe

AliSoltanianFJ committed:
Merge pull request #48 from AliSoltanianFJ/FixRetryOptionsConstructorRecipeBug

Fixed bug that crashed the RetryOptionsConstructorRecipe

AliSoltanianFJ committed:
Add unit tests for miscellaneous annotation types

AliSoltanianFJ committed:
Merge pull request #49 from AliSoltanianFJ/AddMiscAnnotationsUnitTests

Add unit tests for miscellaneous annotation types

AliSoltanianFJ committed:
Added recipe to convert TypeReference to ParameterizedType

AliSoltanianFJ committed:
Merge pull request #51 from AliSoltanianFJ/Issue28

Added recipe to convert TypeReference to ParameterizedType

AliSoltanianFJ committed:
Added recipes to convert UnexpectedResponseExceptionType and fixed a bug

AliSoltanianFJ committed:
Merge pull request #57 from AliSoltanianFJ/UnexpectedResponseExceptionRecipes

Added recipes to convert UnexpectedResponseExceptionType and fixed a bug

AliSoltanianFJ committed:
Fixed bug in TypeReferenceRecipe

AliSoltanianFJ committed:
Merge pull request #58 from AliSoltanianFJ/FixTypeReferenceRecipeBug

Fixed bug in TypeReferenceRecipe

stoatus-quo committed:
Test added

stoatus-quo committed:
Recipe added, test passed

stoatus-quo committed:
Merge pull request #59 from AliSoltanianFJ/Issue-36-utilConfigurationType

Issue 36 util configuration type

Change com.azure.core.util.Configuration to io.clientcore.core.util.configuration.Configuration

stoatus-quo committed:
Tests added

stoatus-quo committed:
Tests refactored

stoatus-quo committed:
Recipes added -  UnexpectedResponseExceptionTypeTest failing

stoatus-quo committed:
recipe order changed - all tests passing

stoatus-quo committed:
Merge pull request #60 from AliSoltanianFJ/Issue-56-ExceptionTypes

Issue 56 exception types

AliSoltanianFJ committed:
Added recipe to migrate ClientLogger

AliSoltanianFJ committed:
Merge pull request #61 from AliSoltanianFJ/Issue55ClientLogger

Added recipe to migrate ClientLogger

AliSoltanianFJ committed:
Added recipe to convert BinaryData type

AliSoltanianFJ committed:
Merge pull request #62 from AliSoltanianFJ/AddBinaryDataRecipe

Added recipe to convert BinaryData type

AliSoltanianFJ committed:
Added CI

srnagar commented on rewrite-java-core/src/main/resources/META-INF/rewrite/rewrite.yml:
```diff
@@ -10,7 +10,30 @@ recipeList:
   #
+  # ----------------------------------
+
+  # Recipes that converts the parameters and type of the @UnexpectedResponseExceptionType annotation
```
We don't need to convert annotations. Users don't need to use the annotations directly. Annotations are only needed by the Azure libraries. So, we don't need to update them using OpenRewrite recipes.

AliSoltanianFJ replied:
Removed recipes that convert annotations and their associated unit tests in commit [5447e27](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/5447e27b96a32983846088fe796effd5c4256353)

srnagar commented on rewrite-java-core/src/test/java/BinaryDataTest.java:
```diff
@@ -0,0 +1,42 @@
+
+    /* Test to make sure BinaryData type and import is changed */
+    @Test
+    public void testClientLoggerTypeAndImportChanged() {
```
We should also add tests that use TypeReference in BinaryData.

AliSoltanianFJ replied:
Added unit test that tests BinaryData change with TypeReference in commit [5447e27](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/5447e27b96a32983846088fe796effd5c4256353)

srnagar commented on rewrite-sample/azure-ai-translation-text-v2/libs/com/azure/azure-core-v2/1.0.0-beta.1/_remote.repositories:
```diff
@@ -0,0 +1,8 @@
+#NOTE: This is a Maven Resolver internal implementation file, its format can be changed without prior notice.
```
We should not check in this `libs` directory.

AliSoltanianFJ replied:
Removed libs directory in commit [5447e27](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/5447e27b96a32983846088fe796effd5c4256353)

AliSoltanianFJ committed:
Addressed feedback for 3 Sep 2024 PR

srnagar commented on .github/workflows/maven.yml:
```diff
@@ -0,0 +1,27 @@
+      - name: Clean install
+        run: mvn clean install
+      
+      - name: Execute Unit Tests
```
Do we need this extra step? `mvn clean install` in the previous step also runs the unit tests.

AliSoltanianFJ replied:
Removed in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

srnagar commented on rewrite-java-core/src/main/java/com/azure/recipes/v2recipes/TypeReferenceRecipe.java:
```diff
@@ -0,0 +1,125 @@
+
+    @Override
+    public @NotNull String getDescription() {
+        return "This recipe converts TypeReference<List<T>> to ParameterizedType and removes the import statement for TypeReference.";
```
This should replace not just `List<T>`, it should replace any generic types like `Map<T,V>`, `Set<T>`, `ArrayList<T>` and even some custom generic types like `Foo<T>`. So, this recipe has to be handle all generic types.

Also, `TypeReference` can be used on non-generic classes too and those have to be converted to `Type` as well.

AliSoltanianFJ replied:
Added functionality to replace other generic and non-generic types for TypeReference in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

srnagar commented on rewrite-java-core/src/main/resources/META-INF/rewrite/rewrite.yml:
```diff
@@ -21,4 +27,58 @@ recipeList:
+  # TODO: Filter for implementing classes
+  - org.openrewrite.java.ChangeMethodName:
+      methodPattern: com.azure.core.client.traits.HttpTrait retryOptions(..)
+      newMethodName: httpRetryOptions
```
Doesn't this need to update the package name as well?

AliSoltanianFJ replied:
The HttpTrait recipes are currently unfinished and weren't supposed to be included in this PR (were included by mistake). Removed in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

stoatus-quo replied:
Package name will change for all HttpTrait methods when HttpTrait is changed. This is safer to change as the last thing.

srnagar commented on rewrite-java-core/src/test/java/BinaryDataTest.java:
```diff
@@ -0,0 +1,92 @@
+import com.azure.recipes.v2recipes.TypeReferenceRecipe;
```
These tests should also be in the root package  - `com.azure.recipes.core.v2`. (See my comment above about changing the root package name)

AliSoltanianFJ replied:
Changed package name in commit [8e6d539](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/8e6d539b5b0576d7b7e17a656beac7f15941b3c8)

srnagar commented on rewrite-java-core/src/main/java/com/azure/recipes/v2recipes/TypeReferenceRecipe.java:
```diff
@@ -0,0 +1,125 @@
+package com.azure.recipes.v2recipes;
```
Instead of `v2recipes` we should call this `core.v2` i.e `com.azure.recipes.core.v2`

AliSoltanianFJ replied:
Changed in commit [8e6d539](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/8e6d539b5b0576d7b7e17a656beac7f15941b3c8)

srnagar commented on rewrite-java-core/src/test/java/BinaryDataTest.java:
```diff
@@ -0,0 +1,92 @@
+        @Language("java") String before = "import com.azure.core.util.BinaryData;";
+        before += "\npublic class Testing {";
+        before += "\n  public Testing(){";
+        before += "\n    com.azure.core.util.BinaryData b = BinaryData.fromObject(null);";
```
We have an import above. So, do we need to use fully qualified name here?

AliSoltanianFJ replied:
Fixed in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

srnagar commented on rewrite-java-core/src/test/java/CredentialTest.java:
```diff
@@ -0,0 +1,48 @@
+        @Language("java") String before = "import com.azure.core.credential.KeyCredential;";
+        before += "\npublic class Testing {";
+        before += "\n  public Testing(){";
+        before += "\n    com.azure.core.credential.KeyCredential kc = new KeyCredential(\"<api-key>\");";
```
Same here - we should not use fully qualified names when there's an import. Also, we should add another test where there's not import and the fully qualified name is used.

AliSoltanianFJ replied:
Done in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

srnagar commented on rewrite-java-core/src/test/java/ExceptionTypesTest.java:
```diff
@@ -0,0 +1,148 @@
+        @Language("java") String after = "import com.azure.core.v2.exception.ClientAuthenticationException;";
+        after += "\n\npublic class Testing {";
+        after += "\n  public Testing(){";
+        after += "\n    com.azure.core.v2.exception.ClientAuthenticationException e = new ClientAuthenticationException(null,null);";
```
Users generally don't new up exception types. They just handle these exceptions in a try-catch block. So, we should have tests for converting the exceptions in `catch` block.

AliSoltanianFJ replied:
Modified exception tests to test out exceptions in catch blocks in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

srnagar commented on rewrite-java-core/src/test/java/HttpTraitTest.java:
```diff
@@ -0,0 +1,138 @@
+                "import com.azure.core.http.policy.RetryOptions;\n" +
+                "import com.azure.core.util.ClientOptions;\n" +
+                "\n" +
+                "public class BlankHttpTrait implements HttpTrait<BlankHttpTrait> {\n" +
```
Users generally don't implement traits. We should test how the builder code is transformed by the recipes. For e.g. `clientBuilder.addPolicy(new HttpPipelinePolicy(){...})` is transformed to `clientBuilder.addHttpPipelinePolicy(new HttpPipelinePolicy(){...})`

AliSoltanianFJ replied:
The HttpTrait recipes are currently unfinished and weren't supposed to be included in this PR (were included by mistake). Removed in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

srnagar commented on rewrite-java-core/src/test/java/RetryOptionsConstructorTest.java:
```diff
@@ -56,16 +56,20 @@ void testChangeRetryOptionsTypeNoArgInit() {
         @Language("java") String before = "import com.azure.core.http.policy.RetryOptions;import java.time.Duration;import com.azure.core.http.policy.FixedDelayOptions;";
         before += "\npublic class Testing {";
         before += "\n  FixedDelayOptions f = new FixedDelayOptions(3, Duration.ofMillis(50));";
+        before += "\n  String test = \"testing\";";
```
Why is this needed?

AliSoltanianFJ replied:
Removed in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

srnagar commented on rewrite-java-core/src/test/java/TypeReferenceTest.java:
```diff
@@ -0,0 +1,67 @@
+     * @param spec stores settings for testing environment; e.g. which recipes to use for testing
+     */
+    @Override
+    public void defaults(RecipeSpec spec) {
```
We should import all recipes to ensure we test how all recipes interact with each other. There might be some overlap that might cause some overwriting to happen and result in unexpected transformation. So, I would suggest having a base test class where all recipes are added in RecipeSpec and all other tests inheriting from it.

AliSoltanianFJ replied:
Ensured all testing classes only use `rewrite.yml`. Moved all imperative (Java-based) recipes into `rewrite.yml`.

See changes in commit [492daeb](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/492daeb48360470829cc54218911b9b3e1758ace)

AliSoltanianFJ committed:
Address Feedback - 1

AliSoltanianFJ committed:
Addressed Feedback - 2

srnagar commented on .gitignore:
```diff
@@ -410,7 +410,6 @@ FodyWeavers.xsd
 .mtj.tmp/
 
 # Package Files #
-*.jar
```
nit: we should not check-in jar files

AliSoltanianFJ replied:
Addressed in commit [a29f16e](https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14/commits/a29f16e6f730a6332908c520c0bd5244ffe4406c), should be ready to merge

srnagar reviewed (approved):
Looks good to me

AliSoltanianFJ committed:
Fixed nit comment: .gitignore changed to ignore jars

samvaity commented on rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java:
```diff
@@ -1,4 +1,4 @@
+package com.azure.recipes.v2;
 
 import org.jetbrains.annotations.NotNull;
 import org.openrewrite.*;
```
nit: we should avoid using star imports.

AliSoltanianFJ replied:
Addressed in latest commits.

samvaity commented on rewrite-java-core/src/test/java/com/azure/recipes/v2/TypeReferenceTest.java:
```diff
@@ -0,0 +1,148 @@
+        before += "\nimport java.lang.reflect.Type;";
+        before += "\nimport com.azure.core.util.serializer.TypeReference;";
+        before += "\npublic class Testing {";
+        before += "\n  private static final TypeReference<java.util.Map<String,Integer>> TESTING_TYPE = new TypeReference<java.util.Map<String, Integer>>() {\n  };";
```
```suggestion
before += "\n  private static final TypeReference<java.util.Map<String, Integer>> TESTING_TYPE = new TypeReference<java.util.Map<String, Integer>>() {\n  };";
```

samvaity commented on rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java:
```diff
@@ -47,15 +47,30 @@ private static class ConvertTypeReferenceVisitor extends JavaIsoVisitor<Executio
-                // Extract type from List type in TypeReference declaration
-                Pattern pattern = Pattern.compile("List<(\\w+)>");
+                // Extract type from generic type in TypeReference declaration
+                Pattern pattern = Pattern.compile("<([\\w.]+)<([\\w.]+)(?:,\\s*([\\w.]+))?>>");
```
Instead of parsing the string representation of the class, use the structured metadata provided by `J.NewClass`
here https://github.com/openrewrite/rewrite/blob/main/rewrite-java/src/main/java/org/openrewrite/java/tree/J.java
This will help in giving you access to its type arguments, same way without the regex.

AliSoltanianFJ replied:
Addressed in latest commits.

samvaity commented on rewrite-java-core/src/main/java/com/azure/recipes/v2/TypeReferenceRecipe.java:
```diff
@@ -47,15 +47,30 @@ private static class ConvertTypeReferenceVisitor extends JavaIsoVisitor<Executio
                     .anyMatch(methodDeclaration -> methodDeclaration.getName().getSimpleName().equals("getRawType"));
 
             if (!alreadyTransformed && visitedNewClass.toString().contains("TypeReference")) {
```
Same suggestion here the inferred type can be used here instead of string comparison to `TypeReference`.

AliSoltanianFJ replied:
Addressed in latest commits.

samvaity commented on rewrite-java-core/src/test/java/com/azure/recipes/core/v2/TypeReferenceTest.java:
```diff
@@ -0,0 +1,148 @@
+                "  private static final java.lang.reflect.Type TESTING_TYPE = new ParameterizedType() {\n" +
+                "      @Override\n" +
+                "      public java.lang.reflect.Type getRawType() {\n" +
+                "          return java.util.List.class;\n" +
```
We are still using fully qualified names here and other places.

Earlier comment https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14#discussion_r1750823950

AliSoltanianFJ replied:
Addressed in latest commits. Fully qualified names no longer used when imports are already added.

samvaity commented on rewrite-java-core/src/main/java/com/azure/recipes/core/v2/TypeReferenceRecipe.java:
```diff
@@ -0,0 +1,154 @@
+
+            if (!alreadyTransformed && visitedNewClass.toString().contains("TypeReference")) {
+                // Extract type from generic type in TypeReference declaration
+                Pattern pattern = Pattern.compile("<([\\w.]+)<([\\w.]+)(?:,\\s*([\\w.]+))?>>");
```
https://github.com/Azure/azsdk-java-rewrite-recipes/pull/14#discussion_r1771947112

J.ParameterizedType will give you access to the arguments instead of Regex matcher here.

samvaity commented on rewrite-java-core/src/main/java/com/azure/recipes/core/v2/TypeReferenceRecipe.java:
```diff
@@ -0,0 +1,154 @@
+        @Override
+        public J.@NotNull Import visitImport(J.@NotNull Import importStmt, @NotNull ExecutionContext ctx) {
+            if (importStmt.getQualid().toString().equals("java.lang.reflect.ParameterizedType")) {
+                count ++;
```
Instead of using the count variable for tracking imports (which can be error-prone), you could keep track of the imports in a set.

samvaity commented on rewrite-java-core/src/main/java/com/azure/recipes/core/v2/TypeReferenceRecipe.java:
```diff
@@ -0,0 +1,154 @@
+                    .map(J.MethodDeclaration.class::cast)
+                    .anyMatch(methodDeclaration -> methodDeclaration.getName().getSimpleName().equals("getRawType"));
+
+            if (!alreadyTransformed && visitedNewClass.toString().contains("TypeReference")) {
```
Could use if (visitedNewClass.getClazz() instanceof J.ParameterizedType) {
?

AliSoltanianFJ committed:
Updated TypeReference recipe to use OpenRewrite type info

AliSoltanianFJ committed:
Remove fully qualified names when using import

AliSoltanianFJ committed:
Added comments to explain changes

AliSoltanianFJ committed:
Extract type info from getClazz

AliSoltanianFJ committed:
Remove more fully qualified names

AliSoltanianFJ committed:
Fixed bug in TypeReference recipe where import for java.lang.reflect.Type is not added

AliSoltanianFJ commented:
Changes according to feedback/reviews:
- Ensured fully qualified types are not used whenever an import is already added
- Extracted type information in TypeReferenceRecipe where possible
- Fixed a bug in TypeReference recipe that would not add the required import for `java.lang.reflect.Type`
- Removed type validation from TypeReference test due to bug in OpenRewrite test framework parser. Note that the tests still pass as the transformed code matches expected output.

AliSoltanianFJ commented:
Changes according to feedback/reviews:
- Ensured fully qualified types are not used in unit testing code whenever an import is already added
- Extracted type information in TypeReferenceRecipe where possible
- Fixed a bug in TypeReference recipe that would not add the required import for `java.lang.reflect.Type`
- Fixed a bug in RetryOptionsConstructor recipe which would cause it to crash when it parsed certain Java types while searching for RetryOptions
- Removed type validation from TypeReference test due to bug in OpenRewrite test framework parser that would not allow it to parse the instantiation of TypeReference. Note that the tests still pass as the transformed code matches expected output.
- Removed usages of star (*) imports

AliSoltanianFJ committed:
Changed to use Set to keep track of import count rather than counter
